### ISSUE TITLE
Prisma 2.20.x increased archive size exceeds the maximum size limit of 50mb for lambda functions on Vercel

### ISSUE THREAD

**Issue by styxlab** [prisma/prisma#7016]
### Bug description

After upgrading prisma packages from 2.19.0 to 2.20.x I get the following message when deploying my Next.js project on Vercel:

```
Failed to process build result for "api/v1/graphql". Data: {"type":"Lambda","zipBuffer":{"type":"Buffer","data":[80,75,3,4,20,0,0,8,8,0,193,174,166,82,203,126,12,106,139,71,0,0,126,92,10,0,36,0,0,0,110,111,100,101,95,109,111,100,117,108,101,115,47,46,112,114,105,115,109,97,47,99,108,105,101,110,116,47,105,110,100,.

Error: The Serverless Function "api/core/v1/graphql" is 78.49mb which exceeds the maximum size limit of 50mb. Learn More: https://vercel.link/serverless-function-size

```
with
```


23:54:03.286 | Serverless Function's page: api/v1/graphql.js
-- | --
23:54:03.291 | Large Dependencies                          Uncompressed size  Compressed size
23:54:03.291 | node_modules/@prisma/engines                           113 MB          39.6 MB
23:54:03.291 | node_modules/.prisma/client                           46.2 MB          16.1 MB
23:54:03.292 | node_modules/prisma/build                             5.27 MB          1.09 MB
23:54:03.292 | node_modules/@prisma/client                           3.93 MB           822 kB
23:54:03.292 | node_modules/prettier/parser-typescript.js            3.07 MB           788 kB
23:54:03.292 | node_modules/highlight.js/lib                         1.41 MB           458 kB
23:54:03.292 | node_modules/prettier/index.js                        1.66 MB           382 kB
23:54:03.292 | node_modules/prettier/parser-flow.js                  3.09 MB           305 kB
23:54:03.292 | node_modules/typeorm/driver                            1.6 MB           214 kB
23:54:03.292 | node_modules/busboy/deps                               618 kB           196 kB
23:54:03.292 | node_modules/micro/node_modules                        360 kB           190 kB
23:54:03.292 | node_modules/encoding/node_modules                     328 kB           179 kB
23:54:03.292 | node_modules/lodash/lodash.js                          544 kB          96.4 kB
23:54:03.292 | All dependencies                                       190 MB          62.6 MB

```

When I revert back to 2.19.0 deploys succeed. Unfortunately, I don't get a report on the archive sizes in the succeeding case.

### How to reproduce

Don't have a reprducable example that I can share at the moment, but I hope the attached information is sufficient.

### Expected behavior

Prisma bundle size should not increase to the point where one hits the AWS limits for serverless functions.

### Prisma information

Here are all prisma related packages in my `package.json` for the failing case (when I change back to 2.19.0 in the two prisma instances below, everything works as expected).

```
  "dependencies": {
    "@nexus/schema": "^0.20.1",
    "@prisma/client": "^2.22.1",
    "@urql/exchange-graphcache": "^4.0.0",
    "apollo-server-micro": "^2.24.0",
    "graphql": "^15.5.0",
    "graphql-tag": "^2.12.4",
    "nexus": "^1.0.0",
    "nexus-plugin-prisma": "^0.34.1",
    "urql": "^2.0.2",
  },
  "devDependencies": {
    "@graphql-codegen/cli": "^1.21.4",
    "@graphql-codegen/typescript": "^1.22.0",
    "@graphql-codegen/typescript-operations": "^1.17.16",
    "@graphql-codegen/typescript-urql": "^2.0.6",
    "next-prisma-plugin": "^0.1.1",
    "prisma": "^2.22.1",
  }
```

### Environment & setup

- OS: Fedora
- Database: PostgreSQL
- Node.js version: 14.x
- Prisma version: 2.22.1 (failing for 2.20.x)

**Comment by styxlab** [comment]
@Jolg42 Thanks so much for looking into the issue. Yes, the package.json only contains packages which I thought are relevant. I definitely don't have the `prisma` CLI in there and recreated my lock file many times, so I doubt this is the issue.

Unfortunately, I cannot share the entire project but will try to make an example. Totally understand that you need something to work with.

**Comment by janpio** [comment]
Hi @styxlab, can you quickly try the latest versions of Prisma CLI and Prisma Client in your project and report back if this already fixes the problem (both `latest` and `dev` if possible - you can ignore warnings from `nexus-plugin-prisma` in this case)? There was a known problem in the interplay between Prisma and Vercel tooling that lead to _some_ problem that could cause what you are seeing.

The fundamental issue here is that your Vercel archive includes the `@prisma/engines` which it should not. This package is part of the Prisma CLI (which it also includes as `node_modules/prisma`) and should not be included by Vercel when deploying.

If the above does not solve the problem already, we will need to look deeper into your project to pinpoint the problem and understand better what is going on.

(We are also looking in workarounds that fix it more directly - but have not been successful yet.)

**Comment by Jolg42** [comment]
Likely a duplicate of #6932

**Comment by janpio** [comment]
Thanks so much for the reproduction. 

I would indeed say this is related to the same root cause as https://github.com/prisma/prisma/issues/6932 and kind of a duplicate of that issue: Nexus and Nexus Plugin Prisma together with Prisma Client and CLI somehow have a weird interaction, which leads to Vercel's `@vercel/nft` (which decides what needs to be included in the package and what does not) including `@prisma/engines` in the archive leading to the error you are getting.

After I looked into the reproduction project, I also see that you are indeed using Nextjs in your `package.json` as well as the other project.

The next release next week Tuesday will include the fix from `dev` and should fix the problem then and unblock you.

**Comment by styxlab** [comment]
Sounds good! 

Do you know how we could verify that `@prisma/engines` is not included with the latest prisma `dev` version when deployed to Vercel? Would be nice to see the total archive size (zipped and unzipped), but logs only show this info in the error case.

**Comment by styxlab** [comment]
Thanks for the explanations. Love Prisma!

**Comment by janpio** [comment]
I just learned a bit more (from Vercel): 

- `@vercel/nft` is actually a CLI. After you built your project, you can also use `yarn nft print ./.next/server/pages/api/core/v1/graphql.js` to output the list of included files for this one serverless function.
- And with `yarn nft build ./.next/server/pages/api/core/v1/graphql.js` you can build the same function archive that Vercel builds in their build process as well, and `du -sh dist` then outputs the archive size.
- The only thing that is now still missing is the function that takes that archive and outputs it sorted by package size, and shows both compressed and uncompressed size. That unfortunately is not open source (yet).

**Comment by rubenkaiser** [comment]
> I managed to strip down my project to a minimal example. It could probably be reduced even further, but I think it is a good starting point for a proper analysis. Here are the steps to reproduce:
> 
> 1. `git clone https://github.com/styxlab/vercel-prisma-error.git`
> 2. `yarn; yarn build` (just test it is compiling without errors)
> 3. `vercel --prod` (deploy with vercel CLI)
> 4. See the deployment fail
> 5. Inspect the logs with `vercel logs <*.vercel.app>`
> 
> Error should have the following pattern
> 
> ```
> 2021-05-11T13:36:45.278Z  Traced Next.js server files in: 2.640s
> 2021-05-11T13:37:29.021Z  Warning: Max serverless function size of 50 MB compressed or 250 MB uncompressed hit
> 2021-05-11T13:37:29.022Z  Serverless Function's page: api/core/v1/graphql.js
> 2021-05-11T13:37:29.026Z  Large Dependencies                  Uncompressed size  Compressed size
> 2021-05-11T13:37:29.027Z  node_modules/@prisma/engines                   113 MB          39.6 MB
> 2021-05-11T13:37:29.027Z  node_modules/.prisma/client                   45.4 MB          16.1 MB
> 2021-05-11T13:37:29.027Z  node_modules/prisma/build                      5.3 MB          1.09 MB
> 2021-05-11T13:37:29.027Z  node_modules/@prisma/client                   3.93 MB           823 kB
> 2021-05-11T13:37:29.027Z  node_modules/busboy/deps                       618 kB           196 kB
> 2021-05-11T13:37:29.027Z  node_modules/micro/node_modules                360 kB           190 kB
> 2021-05-11T13:37:29.027Z  node_modules/encoding/node_modules             328 kB           179 kB
> 2021-05-11T13:37:29.027Z  node_modules/lodash/lodash.js                  544 kB          96.4 kB
> 2021-05-11T13:37:29.027Z  
> 2021-05-11T13:37:29.027Z  All dependencies                               173 MB          59.1 MB
> 2021-05-11T13:37:29.028Z  Max serverless function size was exceeded for 1 function
> 2021-05-11T13:37:29.202Z  Created all serverless functions in: 43.924s
> 2021-05-11T13:37:43.370Z  Failed to process build result for "api/core/v1/graphql". Data: {"type":"Lambda","zipBuffer":{"type":"Buffer","data":[80,75,3,4,20,0,0,8,8,0,174,108,171,82,104,52,206,48,15,6,0,0,57,24,0,0,31,0,0,0,46,110,101,120,116,47,115,101,114,118,101,114,47,119,101,98,112,97,99,107,45,114,117,110,116,105,109,101,46,106,115,173,88.
> 2021-05-11T13:37:43.377Z  Error: The Serverless Function "api/core/v1/graphql" is 57.31mb which exceeds the maximum size limit of 50mb. Learn More: https://vercel.link/serverless-function-size
> ```
> 
> Let me know if you need anything else.



> I managed to strip down my project to a minimal example. It could probably be reduced even further, but I think it is a good starting point for a proper analysis. Here are the steps to reproduce:
> 
> 1. `git clone https://github.com/styxlab/vercel-prisma-error.git`
> 2. `yarn; yarn build` (just test it is compiling without errors)
> 3. `vercel --prod` (deploy with vercel CLI)
> 4. See the deployment fail
> 5. Inspect the logs with `vercel logs <*.vercel.app>`
> 
> Error should have the following pattern
> 
> ```
> 2021-05-11T13:36:45.278Z  Traced Next.js server files in: 2.640s
> 2021-05-11T13:37:29.021Z  Warning: Max serverless function size of 50 MB compressed or 250 MB uncompressed hit
> 2021-05-11T13:37:29.022Z  Serverless Function's page: api/core/v1/graphql.js
> 2021-05-11T13:37:29.026Z  Large Dependencies                  Uncompressed size  Compressed size
> 2021-05-11T13:37:29.027Z  node_modules/@prisma/engines                   113 MB          39.6 MB
> 2021-05-11T13:37:29.027Z  node_modules/.prisma/client                   45.4 MB          16.1 MB
> 2021-05-11T13:37:29.027Z  node_modules/prisma/build                      5.3 MB          1.09 MB
> 2021-05-11T13:37:29.027Z  node_modules/@prisma/client                   3.93 MB           823 kB
> 2021-05-11T13:37:29.027Z  node_modules/busboy/deps                       618 kB           196 kB
> 2021-05-11T13:37:29.027Z  node_modules/micro/node_modules                360 kB           190 kB
> 2021-05-11T13:37:29.027Z  node_modules/encoding/node_modules             328 kB           179 kB
> 2021-05-11T13:37:29.027Z  node_modules/lodash/lodash.js                  544 kB          96.4 kB
> 2021-05-11T13:37:29.027Z  
> 2021-05-11T13:37:29.027Z  All dependencies                               173 MB          59.1 MB
> 2021-05-11T13:37:29.028Z  Max serverless function size was exceeded for 1 function
> 2021-05-11T13:37:29.202Z  Created all serverless functions in: 43.924s
> 2021-05-11T13:37:43.370Z  Failed to process build result for "api/core/v1/graphql". Data: {"type":"Lambda","zipBuffer":{"type":"Buffer","data":[80,75,3,4,20,0,0,8,8,0,174,108,171,82,104,52,206,48,15,6,0,0,57,24,0,0,31,0,0,0,46,110,101,120,116,47,115,101,114,118,101,114,47,119,101,98,112,97,99,107,45,114,117,110,116,105,109,101,46,106,115,173,88.
> 2021-05-11T13:37:43.377Z  Error: The Serverless Function "api/core/v1/graphql" is 57.31mb which exceeds the maximum size limit of 50mb. Learn More: https://vercel.link/serverless-function-size
> ```
> 
> Let me know if you need anything else.

@styxlab did you do something to get this output? I'm only getting two lines, as opposed to your output:
```
2021-05-12T07:05:38.640Z  Traced Next.js server files in: 31.092s
2021-05-12T07:06:04.620Z  Created all serverless functions in: 25.979s
```

**Comment by janpio** [comment]
And final update after feedback from Vercel: 
- You should be able to set `NEXT_DEBUG_FUNCTION_SIZE =1` on your Vercel app to get more output during the build, even when the serverless function does _not_ fail to upload. This is what it looks for a project that _just_ uses `@prisma/client` to run some simple queries:


	  2021-05-11T23:26:09.207Z  Serverless function size info
	  2021-05-11T23:26:09.207Z  Serverless Function's page: api.js
	  2021-05-11T23:26:09.207Z  Large Dependencies                  Uncompressed size  Compressed size
	  2021-05-11T23:26:09.207Z  node_modules/.prisma/client                   45.7 MB          16.2 MB
	  2021-05-11T23:26:09.207Z  node_modules/@prisma/client                   1.17 MB           243 kB
	  2021-05-11T23:26:09.207Z  node_modules/encoding/node_modules             328 kB           179 kB
	  2021-05-11T23:26:09.208Z  
	  2021-05-11T23:26:09.208Z  All dependencies                              47.2 MB          16.6 MB
	  2021-05-11T23:26:09.251Z  Created all serverless functions in: 8.564s
	  
  Neat!

**Comment by styxlab** [comment]
That new Vercel flag is **exactly** what I have been looking for! For completeness, here is the output after upgrading to latest `dev` version for the reproduction example I provided earlier.
```12:06:44.042 | Serverless function size info
12:06:44.042 | Serverless Function's page: api/core/v1/graphql.js
12:06:44.046 | Large Dependencies                  Uncompressed size  Compressed size
12:06:44.047 | node_modules/.prisma/client                   45.8 MB          16.2 MB
12:06:44.047 | node_modules/prisma/build                     4.17 MB           920 kB
12:06:44.047 | node_modules/@prisma/client                   3.94 MB           823 kB
12:06:44.047 | node_modules/busboy/deps                       618 kB           196 kB
12:06:44.047 | node_modules/micro/node_modules                360 kB           190 kB
12:06:44.047 | node_modules/encoding/node_modules             328 kB           179 kB
12:06:44.047 | node_modules/@prisma/engines                   597 kB           123 kB
12:06:44.047 | node_modules/lodash/lodash.js                  544 kB          96.4 kB
12:06:44.047 | All dependencies                              59.8 MB          19.6 MB
12:06:44.149 | Created all serverless functions in: 17.550s
```

**Comment by janpio** [comment]
Final update in the other thread, as we could figure out the _reason_ why all this was happening: https://github.com/prisma/prisma/issues/6932#issuecomment-843539427
